### PR TITLE
Add Driver::pread_exact method.

### PR DESCRIPTION
**Pull request by doriath**
This is similar to std::io::Read::read_exact, that reads the exact
number of bytes to fill in the input buffers. The current code often
assumes that the buffer will be fully filled in (especially when reading
u32 into a 4 byte buffer), this method will make it much easier to write
it correctly.

### CODE REVIEW

File: src/drivers/model/src/lib.rs [hunk 1]
```diff
@@ -20,4 +19,16 @@ pub trait Driver {
     fn pwrite(&mut self, data: &[u8], pos: usize) -> Result<usize>;
     /// Shutdown the device.
     fn shutdown(&mut self);
+    /// Reads the exact number of bytes to fill in the `data`.
+    /// Returns ok if `data` is empty.
+    fn pread_exact(&self, data: &mut [u8], pos: usize) -> Result<()> {
+        let mut r = 0;
+        while r < data.len() {
```
Comment (fotonick) [inline src/drivers/model/src/lib.rs:26]: Is it possible to read more data than you asked for?

File: src/drivers/model/src/lib.rs [hunk 2]
```diff
@@ -20,4 +19,16 @@ pub trait Driver {
...
+            r += self.pread(&mut data[r..], pos + r)?;
```
Comment (fotonick) [inline src/drivers/model/src/lib.rs:27]: Do we want to guard against an infinite loop of reading 0 bytes?

**Review by fotonick — COMMENTED** [review]
I love mirroring the semantics of something well known for readability.

File: src/drivers/model/src/lib.rs (hunk 1 above)
Comment (doriath) [reply src/drivers/model/src/lib.rs]: No, unless the pread implementation is invalid.

File: src/drivers/model/src/lib.rs (hunk 2 above)
Comment (doriath) [reply src/drivers/model/src/lib.rs]: Added protection against reading 0 bytes.

File: src/drivers/model/src/lib.rs (hunk 1 above)
Comment (fotonick) [reply src/drivers/model/src/lib.rs]: Sounds good. Just wanted you to think about it.

File: src/drivers/model/src/lib.rs (hunk 2 above)
Comment (fotonick) [reply src/drivers/model/src/lib.rs]: Thanks. Does zero bytes mean EOF? It doesn't mean time out or some other random failure condition (like C libraries like to alias together)? If our spec doesn't specify, one possibility is that we may wish to constrain the return type of `pread` to `Result<NonZeroUsize>` to force future implementers of `Driver` to give everything a well defined meaning. I might be over-thinking this, but this kind of ambiguity does have a habit of breeding bugs.

**Review by fotonick — APPROVED** [review]

**Pull request merged**